gdal/port/cpl_userfaultfd.cpp

```diff
@@ -132,44 +149,144 @@ static void * fault_handler(void * ptr)
+    // system[1]), (3) unregistering the VMA from userfaultfd,
+    // remapping the VMA to evict the pages, registering the VMA
+    // again, (4) making the VMA accessible again, and finally (5)
+    // restoring the previous signal-handling behavior.
+    //
+    // [1] https://lists.debian.org/debian-bsd/2011/05/msg00032.html
+    pthread_mutex_lock(&mutex);
+    if ((ctx->page_limit > 0) && (ctx->pages_used > ctx->page_limit)) {
+      struct sigaction segv;
+      struct sigaction old_segv;
+      struct sigaction bus;
+      struct sigaction old_bus;
+
+      // Step 1 from the block comment above
+      segv.sa_handler = signal_handler;
+      bus.sa_handler = signal_handler;
+      if (sigaction(SIGSEGV, &segv, &old_segv) == -1) {
+      	CPLError(CE_Failure, CPLE_AppDefined,
+      		 "fault_handler: sigaction(SIGSEGV) failed");
+      	break;
+      }
+      if (sigaction(SIGBUS, &bus, &old_bus) == -1) {
+      	CPLError(CE_Failure, CPLE_AppDefined,
+      		 "fault_handler: sigaction(SIGBUS) failed");
+      	break;
       }
-    }
 
-    // Copy data into page
-    if (VSIFSeekL(file, offset, SEEK_SET)) break;
-    if (VSIFReadL(ctx->page_ptr, bytes_needed, 1, file) != 1) break;
-    pages_used++;
-
-    uffdio_copy.src = reinterpret_cast<uintptr_t>(ctx->page_ptr);
-    uffdio_copy.dst = fault_addr;
-    uffdio_copy.len = static_cast<uintptr_t>(ctx->page_size);
-    uffdio_copy.mode = 0;
-    uffdio_copy.copy = 0;
-    if (ioctl(ctx->uffd, UFFDIO_COPY, &uffdio_copy) == -1) break;
+      // WARNING LACK OF THREAD-SAFETY.
+      //
+      // For example, if a user program (or another part of the
+      // library) installs a SIGSEGV or SIGBUS handler from another
+      // thread after this one has installed its handlers but before
+      // this one uninstalls its handlers, the intervening handler
+      // will be eliminated.  There are other examples, as well, but
+      // there can only be a problems with other threads because the
+      // faulting thread is blocked here.
+      //
+      // This implies that one should not create `/vsimem` handles
+      // while other threads are actively generating faults that use
+      // this mechanism.
+      //
+      // Having multiple active threads that use this mechanism but
+      // with no changes to signal-handling in other threads is NOT a
+      // problem.
+
+      // Step 2
+      if (mprotect(ctx->vma_ptr, ctx->vma_size, PROT_NONE) == -1) {
+      	CPLError(CE_Failure, CPLE_AppDefined,
+      		 "fault_handler: mprotect() failed");
+      	break;
+      }
+
+      // Step 3
+      if (ioctl(ctx->uffd, UFFDIO_UNREGISTER, &ctx->uffdio_register)) {
+      	CPLError(CE_Failure, CPLE_AppDefined,
+      		 "fault_handler: ioctl(UFFDIO_UNREGISTER) failed");
+      	break;
+      }
+      ctx->vma_ptr = mmap(ctx->vma_ptr, ctx->vma_size, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_FIXED, -1, 0);
+      if (ctx->vma_ptr == BAD_MMAP) {
+      	CPLError(CE_Failure, CPLE_AppDefined,
```

Comment:
vma_ptr should likely be nullified for proper behaviour of the cleanup function

Reply:
Done.